core/amber/src/main/scala/edu/uci/ics/texera/web/resource/WorkflowWebsocketResource.scala

```diff
@@ -158,34 +190,247 @@ class WorkflowWebsocketResource {
+    } else if (Constants.storage.equals("mongodb")) {
+      logger.info("use mongodb for materialization")
+      opResultStorage = new MongoOpResultStorage()
+    } else {
+      logger.info("invalid storage config")
+      System.exit(-1)
+    }
+  }
+
   def executeWorkflow(session: Session, request: ExecuteWorkflowRequest): Unit = {
+    var cachedOperators: mutable.HashMap[String, OperatorDescriptor] = null
+    var cacheSourceOperators: mutable.HashMap[String, CacheSourceOpDesc] = null
+    var cacheSinkOperators: mutable.HashMap[String, CacheSinkOpDesc] = null
+    var operatorRecord: mutable.HashMap[String, WorkflowVertex] = null
+    if (opResultSwitch) {
+      if (!sessionCachedOperators.contains(session.getId)) {
+        cachedOperators = mutable.HashMap[String, OperatorDescriptor]()
+        sessionCachedOperators += ((session.getId, cachedOperators))
+      } else {
+        cachedOperators = sessionCachedOperators(session.getId)
+      }
+      if (!sessionCacheSourceOperators.contains(session.getId)) {
+        cacheSourceOperators = mutable.HashMap[String, CacheSourceOpDesc]()
+        sessionCacheSourceOperators += ((session.getId, cacheSourceOperators))
+      } else {
+        cacheSourceOperators = sessionCacheSourceOperators(session.getId)
+      }
+      if (!sessionCacheSinkOperators.contains(session.getId)) {
+        cacheSinkOperators = mutable.HashMap[String, CacheSinkOpDesc]()
+        sessionCacheSinkOperators += ((session.getId, cacheSinkOperators))
+      } else {
+        cacheSinkOperators = sessionCacheSinkOperators(session.getId)
+      }
+      if (!sessionOperatorRecord.contains(session.getId)) {
+        operatorRecord = mutable.HashMap[String, WorkflowVertex]()
+        sessionOperatorRecord += ((session.getId, operatorRecord))
+      } else {
+        operatorRecord = sessionOperatorRecord(session.getId)
+      }
+    }
+
+    logger.info("Session id: {}", session.getId)
     val context = new WorkflowContext
     val jobID = Integer.toString(WorkflowWebsocketResource.nextJobID.incrementAndGet)
     context.jobID = jobID
     context.userID = UserResource
       .getUser(sessionMap(session.getId)._2)
       .map(u => u.getUid)
 
-    val texeraWorkflowCompiler = new WorkflowCompiler(
-      WorkflowInfo(request.operators, request.links, request.breakpoints),
-      context
-    )
+    if (opResultSwitch) {
+      updateCacheStatus(
+        session,
+        CacheStatusUpdateRequest(
+          request.operators,
+          request.links,
+          request.breakpoints,
+          request.cachedOperatorIDs
+        )
+      )
+    }
 
+    var workflowInfo = WorkflowInfo(request.operators, request.links, request.breakpoints)
+    if (opResultSwitch) {
+      workflowInfo.cachedOperatorIDs = request.cachedOperatorIDs
+      logger.info("Cached operators: {}.", cachedOperators.toString())
+      logger.info("request.cachedOperatorIDs: {}.", request.cachedOperatorIDs)
```

Comment:
for those it's better to keep it in debug level.

Reply:
fixed